### PR TITLE
Fixed sometimes failing to read floats and doubles from Packet on Android.

### LINKED ISSUE #1565: Sometimes not able to receive floats via Packet on Android
Thanks for raising your issue here! 🙂

Before you submit the issue however, we'd like you to consider the follow points.

* We like to use the issue tracker for confirmed issues.
* If you're stuck with SFML, please use [the forum](https://en.sfml-dev.org/forums/index.php#c3) to get help.

----

## Subject of the issue

I set up a simple server-client application, the server running on desktop (Ubuntu 18.04) and the client on Android. The server after an initial message from the client sends an `sf::Packet` back to the client, which contains an `std::string` and a `float`. The client crashes when it tries to read the float after reading the string. The error message is as follows:
```2019-02-25 16:53:48.872 25082-25112/org.sfmldev.android A/libc: Fatal signal 7 (SIGBUS), code 1, fault addr 0xea02f07b in tid 25112 (sfmldev.android)```

I think the error comes from some strict requirements about the alignment of `float`s.
See [this related post](https://stackoverflow.com/questions/26158510/gcc-float-pointer-casting-in-c-causing-sigbus-error/26158641#26158641) and the [line in question](https://github.com/SFML/SFML/blob/6d8143088ec8ffc31a12ee3a9a60119474407872/src/SFML/Network/Packet.cpp#L243).
This suspicion is reinforced by the fact that the error does not always happen depending on the length of the string sent.

## Your environment

* Host enviroment: Ubuntu 18.04, client: Android 7.0 Huawei MediaPad M3
* SFML version: git master for the client (Android) and precompiled SFML 2.5.1 for Linux for the server (Ubuntu)
* I built the client (Android) SFML libraries with the following command: `cmake -DANDROID_ABI=armeabi-v7a -DANDROID_PLATFORM=android-21 -stdlib=libc++ -DCMAKE_TOOLCHAIN_FILE=${ndk_path}/build/cmake/android.toolchain.cmake -DCMAKE_ANDROID_NDK_TOOLCHAIN_VERSION=clang -DCMAKE_SYSTEM_NAME=Android -DCMAKE_ANDROID_NDK=${ndk_path} -DCMAKE_ANDROID_STL_TYPE=c++_static -DCMAKE_BUILD_TYPE=Debug ../..
`, where `ndk_path` points to the directory containing the latest NDK version (r19b).

## Steps to reproduce
Server:
```cpp
#include <iostream>
#include <cstdlib>
#include <SFML/Network/UdpSocket.hpp>
#include <SFML/Network/Socket.hpp>
#include <SFML/Network/Packet.hpp>

int main() {
    // Choose an arbitrary port for opening sockets
    const unsigned short port = 50001;

    // Create a socket to receive a message from anyone
    sf::UdpSocket socket;

    // Listen to messages on the specified port
    if (socket.bind(port) != sf::Socket::Done)
        return EXIT_FAILURE;
    std::cout << "Server is listening to port " << port << ", waiting for a message... " << std::endl;

    // Wait for a message
    sf::Packet packet;
    sf::IpAddress sender;
    unsigned short senderPort;
    if (socket.receive(packet, sender, senderPort) != sf::Socket::Done)
        return EXIT_FAILURE;
    std::cout << "Message received from client " << sender << std::endl;

    sf::Packet packet2;
    std::string message = "message";
    packet2 << message;
    float f = 32.234f;
    packet2 << f;
    if (socket.send(packet2, sender, senderPort) != sf::Socket::Done)
        return EXIT_FAILURE;

    std::cout << "Message sent to the client: \"" << message << "\"" << std::endl;
    std::cout << "Float sent to the client: \"" << f << "\"" << std::endl;

    // Wait until the user presses 'enter' key
    std::cout << "Press enter to exit..." << std::endl;
    std::cin.ignore(10000, '\n');
    std::cin.ignore(10000, '\n');

    return EXIT_SUCCESS;
}
```
Client:
```cpp
#include <SFML/System.hpp>
#include <SFML/Window.hpp>
#include <SFML/Graphics.hpp>
#include <SFML/Audio.hpp>
#include <SFML/Network.hpp>
#include <android/log.h>

int main(int argc, char *argv[]) {
    sf::VideoMode screen(sf::VideoMode::getDesktopMode());

    sf::RenderWindow window(screen, "");
    window.setFramerateLimit(30);

    sf::View view = window.getDefaultView();

    sf::Color background = sf::Color::White;

    // We shouldn't try drawing to the screen while in background
    // so we'll have to track that. You can do minor background
    // work, but keep battery life in mind.
    bool active = true;

    while (window.isOpen()) {
        sf::Event event;

        while (active ? window.pollEvent(event) : window.waitEvent(event)) {
            switch (event.type) {
                case sf::Event::Closed:
                    window.close();
                    break;
                case sf::Event::KeyPressed:
                    if (event.key.code == sf::Keyboard::Escape)
                        window.close();
                    break;
                case sf::Event::Resized:
                    view.setSize(event.size.width, event.size.height);
                    view.setCenter(event.size.width / 2, event.size.height / 2);
                    window.setView(view);
                    break;
                case sf::Event::LostFocus:
                    background = sf::Color::Black;
                    break;
                case sf::Event::GainedFocus:
                    background = sf::Color::White;
                    break;

                    // On Android MouseLeft/MouseEntered are (for now) triggered,
                    // whenever the app loses or gains focus.
                case sf::Event::MouseLeft:
                    active = false;
                    break;
                case sf::Event::MouseEntered:
                    active = true;
                    break;
                case sf::Event::TouchBegan:
                    if (event.touch.finger == 0) {
                        // Create a socket for communicating with the server
                        sf::UdpSocket socket{};
                        socket.bind(sf::Socket::AnyPort);

                        // Send a message to the server
                        sf::Packet welcomePacket;
                        std::string welcomeMessage = "Hi, I'm a client";
                        welcomePacket << welcomeMessage;
                        if (socket.send(welcomePacket, /*IP of the server*/, 50001) !=
                            sf::Socket::Done)
                        __android_log_print(ANDROID_LOG_ERROR, "Packet test", "Error while sending message! ");

                        // Receive an answer from anyone (but most likely from the server)
                        sf::Packet packet;
                        sf::IpAddress sender;
                        unsigned short senderPort;
                        if (socket.receive(packet, sender, senderPort) !=
                            sf::Socket::Done)
                            __android_log_print(ANDROID_LOG_ERROR, "Packet test", "Error while receiving message! ");

                        std::string message;
                        float f;
                        packet >> message;
                        packet >> f;
                        __android_log_print(ANDROID_LOG_DEBUG, "Packet test", "Received from server: message: %s, float: %f", message.c_str(), f);
                    }
                    break;
            }
        }

        if (active) {
            window.clear(background);
            window.display();
        } else {
            sf::sleep(sf::milliseconds(100));
        }
    }
}

```
## Expected behavior

The client application should read the `float` from the `sf::Packet`.

## Actual behavior

The client application crashes.

### PR DESCRIPTION
**Pull request by acsbendi**
* [ ] Has this change been discussed on [the forum](https://en.sfml-dev.org/forums/index.php#c3) or in an issue before?
* [X] Does the code follow the SFML [Code Style Guide](https://www.sfml-dev.org/style.php)?
* [X] Have you provided some example/test code for your changes? -> in the issue
* [ ] If you have additional steps which need to be performed list them as tasks!

----

## Description

This PR closes #1565. `reinterpret_cast` has been replaced with `memcopy` so the alignment rules for floats and doubles are not violated.

## Tasks

* [ ] Tested on Linux
* [X] Tested on Windows (server)
* [ ] Tested on macOS
* [ ] Tested on iOS
* [X] Tested on Android (client)

## How to test this PR?

The detailed test case is described in the issue that this PR will close.

### CODE REVIEW

**Comment by binary1248** [comment]
Since you possess a platform on which you can test the alignment behaviour, did you test whether all the other supported data types work as well? Unaligned memory accesses should also cause a SIGBUS to be raised in those cases as well.

Also, this is just personal preference, but I would reword the commit message to indicate that the problem was caused by unaligned memory accesses.

**Comment by acsbendi** [comment]
Yes I've tested it for some other types, such as `sf::Uint64`. It seems only the floating point data types have these strict requirements. 

I think it would be good to include some more information indeed but I think I would need to open a new PR, do you think it would be worth doing that?

**Comment by eXpl0it3r** [comment]
You can also use `git commit --amend` edit the message and `git push origin <your branch> --force-with-lease` to update the PR.

**Comment by binary1248** [comment]
If you ask me, since we're already at it, it would be nice if all the `reinterpret_cast`s were replaced with `memcpy`s. This way we can avoid potential future headaches on platforms with even stricter alignment requirements.

**Comment by acsbendi** [comment]
@eXpl0it3r Thanks for the advice, I managed to amend the commit message in this way :)

**Comment by acsbendi** [comment]
@binary1248 Yes, this is a good idea (also just for the sake of consistency), so I replaced them everywhere in a new commit.

**Comment by binary1248** [comment]
I would squash the commits, the changes are similar enough that they belong together.

File: src/SFML/Network/Packet.cpp [hunk 1]
```diff
@@ -219,7 +224,8 @@ Packet& Packet::operator >>(Uint64& data)
     {
         // Since ntohll is not available everywhere, we have to convert
         // to network byte order (big endian) manually
-        const Uint8* bytes = reinterpret_cast<const Uint8*>(&m_data[m_readPos]);
+        Uint8 bytes[sizeof(data)];
+        std::memcpy(bytes, &m_data[m_readPos], sizeof(data));
```
Comment (dmitry-t) [inline src/SFML/Network/Packet.cpp]: There is no byte order conversion. Is Int64 really serialized not in the network byte order?

File: src/SFML/Network/Packet.cpp [hunk 2]
```diff
@@ -196,7 +200,8 @@ Packet& Packet::operator >>(Int64& data)
     {
         // Since ntohll is not available everywhere, we have to convert
         // to network byte order (big endian) manually
-        const Uint8* bytes = reinterpret_cast<const Uint8*>(&m_data[m_readPos]);
```
Comment (LaurentGomila) [inline src/SFML/Network/Packet.cpp]: Is this memcpy (and the one below) necessary? There should never be alignment issues with uint8.

Comment (acsbendi) [reply src/SFML/Network/Packet.cpp]: They are not. I just thought it's a good idea to be consistent across all reads from `Packet`. Do you think it's better to leave `reinterpret_cast`s here?
